diff -r -u php-5.2.3.orig/ext/standard/mail.c php-5.2.3.mod2/ext/standard/mail.c
--- php-5.2.3.orig/ext/standard/mail.c	2007-03-30 01:28:58.000000000 +0100
+++ php-5.2.3/ext/standard/mail.c	2007-06-25 10:59:59.955165984 +0100
@@ -97,6 +97,7 @@
 	char *force_extra_parameters = INI_STR("mail.force_extra_parameters");
 	char *to_r, *subject_r;
 	char *p, *e;
+	char *headers2=NULL;
 
 	if (PG(safe_mode) && (ZEND_NUM_ARGS() == 5)) {
 		php_error_docref(NULL TSRMLS_CC, E_WARNING, "SAFE MODE Restriction in effect.  The fifth parameter is disabled in SAFE MODE.");
@@ -171,6 +172,58 @@
 		extra_cmd = php_escape_shell_cmd(extra_cmd);
 	}
 
+ 	// add a header in the form
+ 	//	X-PHP-Script: <server_name><php_self> for [<forwarded_for>,]<remote-addr>
+ 	while(1) {
+ 		zval **server, **remote_addr, **forwarded_for, **php_self, **server_name;
+ 		int headers2_len;
+ 
+ 		if (zend_hash_find(&EG(symbol_table), "_SERVER", sizeof("_SERVER"), (void **) &server)==FAILURE)
+ 			break;
+ 		if (Z_TYPE_PP(server)!=IS_ARRAY)
+ 			break;
+ 
+ 		if (zend_hash_find(Z_ARRVAL_PP(server), "REMOTE_ADDR", sizeof("REMOTE_ADDR"), (void **) &remote_addr) == FAILURE)
+ 			break;
+ 
+ 		if (zend_hash_find(Z_ARRVAL_PP(server), "HTTP_X_FORWARDED_FOR", sizeof("HTTP_X_FORWARDED_FOR"), (void **) &forwarded_for) == FAILURE)
+ 			forwarded_for=NULL;
+ 
+ 		if (zend_hash_find(Z_ARRVAL_PP(server), "PHP_SELF", sizeof("PHP_SELF"), (void **) &php_self) == FAILURE)
+ 			break;
+ 
+ 		if (zend_hash_find(Z_ARRVAL_PP(server), "SERVER_NAME", sizeof("SERVER_NAME"), (void **) &server_name) == FAILURE)
+ 			break;
+ 
+ 		headers2_len = 32+Z_STRLEN_PP(server_name)+Z_STRLEN_PP(php_self)+Z_STRLEN_PP(remote_addr);
+ 		if (forwarded_for)
+ 			headers2_len += Z_STRLEN_PP(forwarded_for)+2;
+ 		if (headers)
+ 			headers2_len += headers_len;
+ 
+ 		headers2 = emalloc(headers2_len);
+ 
+ 		strcpy(headers2, "X-PHP-Script: ");
+ 		strcat(headers2, Z_STRVAL_PP(server_name));
+ 		strcat(headers2, Z_STRVAL_PP(php_self));
+ 		strcat(headers2, " for ");
+ 		if (forwarded_for) {
+ 			strcat(headers2, Z_STRVAL_PP(forwarded_for));
+ 			strcat(headers2, ", ");
+ 		}
+ 		strcat(headers2, Z_STRVAL_PP(remote_addr));
+ 
+ 		if (headers) {
+ 			strcat (headers2, "\n");
+ 			strcat (headers2, headers);
+ 		}
+ 
+ 		headers = headers2;
+ 
+ 		break;
+ 	}
+ 
+
 	if (php_mail(to_r, subject_r, message, headers, extra_cmd TSRMLS_CC)) {
 		RETVAL_TRUE;
 	} else {
@@ -186,6 +239,9 @@
 	if (subject_r != subject) {
 		efree(subject_r);
 	}
+	if (headers2) {
+		efree(headers2);
+	}
 }
 /* }}} */
